src/execution/compiler/operator/cte_scan_translator.cpp

```diff
@@ -0,0 +1,245 @@
+      op_(op),
+      col_types_(codegen->NewIdentifier("col_types")),
+      insert_pr_(codegen->NewIdentifier("insert_pr")),
+      read_col_oids_(codegen_->NewIdentifier("read_col_oids")),
+      read_tvi_(codegen_->NewIdentifier("temp_table_iterator")),
+      read_pci_(codegen_->NewIdentifier("read_pci")) {
+  // ToDo(Gautam,Preetansh): Send the complete schema in the plan node.
+  auto &all_columns = op_->GetTableOutputSchema()->GetColumns();
+  for (auto &col : all_columns) {
+    all_types_.emplace_back(static_cast<int>(col.GetType()));
+  }
+
+  std::vector<catalog::Schema::Column> all_schema_columns;
+  for (uint32_t i = 0; i < all_types_.size(); i++) {
+    catalog::Schema::Column col("col" + std::to_string(i + 1), static_cast<type::TypeId>(all_types_[i]), false,
+                                DummyCVE(), static_cast<catalog::col_oid_t>(i + 1));
+    all_schema_columns.push_back(col);
+    col_oids_.push_back(static_cast<catalog::col_oid_t>(i + 1));
+    col_name_to_oid_[all_columns[i].GetName()] = i + 1;
+  }
+
+  // Create the table in the catalog.
+  catalog::Schema schema(all_schema_columns);
+
+  std::vector<uint16_t> attr_sizes;
+  attr_sizes.reserve(storage::NUM_RESERVED_COLUMNS + schema.GetColumns().size());
+
+  for (uint8_t i = 0; i < storage::NUM_RESERVED_COLUMNS; i++) {
+    attr_sizes.emplace_back(8);
+  }
+
+  TERRIER_ASSERT(attr_sizes.size() == storage::NUM_RESERVED_COLUMNS,
+                 "attr_sizes should be initialized with NUM_RESERVED_COLUMNS elements.");
+
+  for (const auto &column : schema.GetColumns()) {
+    attr_sizes.push_back(column.AttrSize());
+  }
+
+  auto offsets = storage::StorageUtil::ComputeBaseAttributeOffsets(attr_sizes, storage::NUM_RESERVED_COLUMNS);
+
+  storage::ColumnMap col_oid_to_id;
+  for (const auto &column : schema.GetColumns()) {
+    switch (column.AttrSize()) {
+      case storage::VARLEN_COLUMN:
+        col_oid_to_id[column.Oid()] = storage::col_id_t(offsets[0]++);
+        break;
+      case 8:
+        col_oid_to_id[column.Oid()] = storage::col_id_t(offsets[1]++);
+        break;
+      case 4:
+        col_oid_to_id[column.Oid()] = storage::col_id_t(offsets[2]++);
+        break;
+      case 2:
+        col_oid_to_id[column.Oid()] = storage::col_id_t(offsets[3]++);
+        break;
+      case 1:
+        col_oid_to_id[column.Oid()] = storage::col_id_t(offsets[4]++);
+        break;
+      default:
+        throw std::runtime_error("unexpected switch case value");
+    }
+  }
+
+  // Use std::map to effectively sort OIDs by their corresponding ID
+  std::map<storage::col_id_t, catalog::col_oid_t> inverse_map;
+
+  // Notice the change in the inverse map argument different from sql_table get projection map function
+  for (auto col_oid : col_oids_) inverse_map[col_oid_to_id[col_oid]] = col_oid;
+
+  // Populate the projection map using the in-order iterator on std::map
+  uint16_t i = 0;
+  for (auto &iter : inverse_map) projection_map_[iter.second] = i++;
```

Comment:
Can we use std::sort here? From my own test of sorting 100,000 std::pair<int, int>, std::sort is 10%~20% faster than using std::map.

Reply:
It doesnt matter as you wont have 100,000 columns in a table. Also this is done to have similar api as SQL Table so that in future when it is replaced by a temp memory space, its easier to change.